### ISSUE TITLE
[RFC] Code Organization Guideline

### ISSUE THREAD

**Issue by Tcc0403** [Kernel-Heim/forge-cute-py#4]
This RFC proposes a code organization guideline to improve readability and scalability.

## Motivation

Currently, it is unclear how we should structure and expose operator interfaces after registering custom kernels.

Below is an analysis of the interfaces we may want to support for each operator.

For a custom kernel, these are the common interfaces, ordered from lowest to highest level:

1. `torch.ops.forge_cute_py.<my_op>`
A low-level interface that enables developers to compose models using `torch.ops` directly, without importing additional Python modules.

2. `torch.autograd.Function`
A PyTorch class that binds the forward and backward passes together. It can be invoked via `<my_autograd_func>.apply()` inside an nn.Module and supports `.backward()` calls.

3. Functional API
Similar to `torch.autograd.Function`, but wrapped in a Python function. Since `torch.autograd.Function` only supports positional arguments, this layer enables keyword arguments, which are less error-prone and improve readability.

4. `torch.nn.Module`
Useful when trainable parameters are involved (e.g., RMSNorm, attention blocks). This interface also provides an easy drop-in replacement when comparing end-to-end performance against other backends.

## Goal

Provide a concrete example of how kernel implementations and their corresponding interfaces should be organized.

### Directory structure
```
forge_cute_py
├── __init__.py
│
├── ...
├── kernels
│   ├── __init__.py        # Exports kernel class
│   └── copy_transpose.py  # Defines the operator class in a quack fashion
│
├── ops
│   ├── __init__.py        # Exports custom operators
│   └── copy_transpose.py  # Defines autograd.Function, functional API, and optional nn.Module
│
├── ref                    # Mirrors the structure of ops/*
│   ├── __init__.py
│   └── copy_transpose.py

```

#### Kernel class in quack style
Methods in a generic CuteDSL kernel class can be grouped into three categories:
1. Initialization
2. Utility functions
3. CuteDSL-related logic

The following example is adapted from
https://github.com/Dao-AILab/quack/blob/main/quack/softmax.py
```python
class Softmax:
    # Category 1: Initialization
    # Typically contains attributes that affect compilation results
    def __init__(
        self,
        dtype: Type[cutlass.Numeric],
        N: int,
        online_softmax: bool = True,
    ):
        ...
    # Category 2: Utility functions used by the kernel
    # Examples: block-size/layout computation, buffer allocation, etc.
    def _threads_per_row(self) -> int:
        N = self.N
        for limit, threads in [
            (64, 8),
            (128, 16),
            (3072, 32),
            (6144, 64),
            (16384, 128),
        ]:
            if N <= limit:
                return threads
        return 256
    # Category 3: CuteDSL-related logic
    # 3.1 Kernel launcher
    # The kernel is JIT-compiled on first invocation and cached for subsequent calls
    @cute.jit
    def __call__(
        self,
        mX: cute.Tensor,
        mO: cute.Tensor,
        stream: cuda.CUstream,
    ):
        # Sanity checks: dtypes, shapes, and dimension compatibility
        assert mX.element_type == self.dtype
        ...

        # Grid / block size computation
        largest_dtype_width = const_expr(
            max(t.element_type.width for t in [mX, mO])
        )
        tiled_copy, tiler_mn, threads_per_row = self._get_tiled_copy(
            vecsize=128 // largest_dtype_width
        )
        num_threads = tiled_copy.size

        # Kernel launch
        self.kernel(  # passing device code arugments
            mX,
            mO,
            tiler_mn,
            tiled_copy,
            threads_per_row,
        ).launch(  # launching with computed gird/block size
            grid=[
                cute.ceil_div(mX.shape[0], tiler_mn[0]),
                self.cluster_n,
                1,
            ],
            block=[num_threads, 1, 1],
            cluster=[1, self.cluster_n, 1]
            if const_expr(self.cluster_n > 1)
            else None,
            stream=stream,
        )

     # 3.2 Device code
     @cute.kernel
     def kernel(
        self,
        mX: cute.Tensor,
        mO: cute.Tensor,
        tiler_mn: cute.Shape,
        tiled_copy: cute.TiledCopy,
        threads_per_row: cutlass.Constexpr[int],
    ):
        ...
```
#### Compilation and cache handling

Ideally, kernels should be cached to avoid repeated compilation and to minimize host-side overhead.

A common pattern is to maintain a compile cache keyed by the kernel’s compilation-relevant parameters (e.g., dtype and problem size). The kernel is compiled lazily on first use and reused for subsequent invocations.

```python
# Define a private operator and register it with torch.ops
@torch.library.custom_op(
    "quack::_softmax_fwd",
    mutates_args={"out"},
)
def _softmax_fwd(x: torch.Tensor, out: torch.Tensor) -> None:
    """Softmax forward pass.

    Args:
        x: Input tensor of shape (M, N)
        out: Output tensor with the same shape as `x`
    """
    # Sanity checks
    assert x.dim() == 2, "Input must be 2D"
    assert x.is_cuda, "Tensor must be on a CUDA device"
    assert x.dtype in (
        torch.float16,
        torch.bfloat16,
        torch.float32,
    ), "Unsupported dtype"

    N = x.size(1)
    dtype, out_dtype = [
        torch2cute_dtype_map[t.dtype] for t in (x, out)
    ]

    # Compilation key captures all parameters that affect code generation
    compile_key = (dtype, out_dtype, N)

    if compile_key not in _softmax_fwd.compile_cache:
        # Symbolic batch dimension for compilation
        batch_sym = cute.sym_int()

        div = math.gcd(128 // dtype.width, N)
        x_cute, out_cute = [
            fake_tensor(dt, (batch_sym, N), div)
            for dt in (dtype, out_dtype)
        ]

        softmax_op = Softmax(dtype, N)

        # JIT-compile and cache the kernel
        _softmax_fwd.compile_cache[compile_key] = cute.compile(
            softmax_op,
            x_cute,
            out_cute,
            cute.runtime.make_fake_stream(
                use_tvm_ffi_env_stream=True
            ),
            options="--enable-tvm-ffi",
        )

    # Invoke the cached kernel
    _softmax_fwd.compile_cache[compile_key](x, out)

# Initialize a compile cache for the kernel
_softmax_fwd.compile_cache = {}
```

TVM-FFI is a powerful tool and should be enabled by default during compilation.
`torch-c-dlpack-ext` is also highly useful, as it allows custom operators to interoperate with Torch tensors out of the box.
See also: https://docs.nvidia.com/cutlass/latest/media/docs/pythonDSL/cute_dsl_general/compile_with_tvm_ffi.html

#### Functional API

Define a Python wrapper to enable keyword arguments and hide output allocation.

```python
def softmax_fwd(x: torch.Tensor) -> torch.Tensor:
    out = torch.empty_like(x)
    _softmax_fwd(x, out)
    return out
```

- Enables keyword-only arguments (unlike torch.autograd.Function)
- Provides a clean, user-facing interface
- Serves as the default entry point for higher-level APIs

Note that this function is forward-only. To support `.backward()` on the output tensor, a `torch.autograd.Function` wrapper is required, which is introduced in the following section.

#### torch.autograd.Function

By implementing the `forward()` and `backward()` static methods using the previously defined functional APIs, `SoftmaxFunction.apply()` produces an output tensor that supports `.backward()`.

```python
class SoftmaxFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = softmax_fwd(x)
        ctx.save_for_backward(y)
        return y

    @staticmethod
    def backward(ctx, dy):
        (y,) = ctx.saved_tensors
        dx = softmax_bwd(dy, y)
        return dx

```

We can wrap the `torch.autograd.Function` in a Python function to expose it as a functional API that also supports .backward() calls.

```python
def softmax(x: torch.Tensor) -> torch.Tensor:
    """Softmax forward pass with automatic differentiation support.

    Args:
        x: Input tensor of shape (M, N)

    Returns:
        Softmax output tensor of same shape as x
    """
    return SoftmaxFunction.apply(x)
```
